fabric_generator/code_generator.py

```diff
@@ -0,0 +1,139 @@
+import abc
+from typing import Literal
+from fabric import Fabric, Tile, Bel
+
+
+class codeGenerator(abc.ABC):
+
+    @property
+    def outFileName(self):
+        return self._outFileName
+
+    @property
+    def content(self):
+        return self._content
+
+    def __init__(self, outFileName):
+        self._outFileName = outFileName
+        self._content = []
+
+    def writeToFile(self):
+        with open(self._outFileName, 'w') as f:
+            f.write("\n".join(self._content))
+
+    def _add(self, line, indentLevel=0) -> None:
+        if indentLevel == 0:
+            self._content.append(line)
+        else:
+            self._content.append(f"{' ':<{4*indentLevel}}" + line)
+
+    def addNewLine(self):
+        self._add("")
+
+    @abc.abstractmethod
+    def addComment(self, comment, onNewLine=False, end="", indentLevel=0) -> None:
+        pass
+
+    @abc.abstractmethod
+    def addHeader(self, name, package='', maxFramesPerCol='', frameBitsPerRow='', ConfigBitMode='FlipFlopChain', indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addHeaderEnd(self, name, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addParameterStart(self, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addParameterEnd(self, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addParameter(self, name, type, value, end=False, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addPortStart(self, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addPortEnd(self, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addPortScalar(self, name, io, end=False, indentLevel=0):
+        pass
+
+    @abc.abstractmethod
+    def addPortVector(self, name, io, width, end=False, indentLevel=0):
```

Comment:
this seems to be a misleading parameter, `width` here is actually `width-1`, i.e. the index of the MSB